### PR TITLE
Align binary math expression streams by time

### LINKED ISSUE #7427: [1.0.2] Continuous query yields incorrect results
**Version tested**: 1.0.2

**Overview**: A CQ calling fields that don't exist in a measurement the first time it is executed, will generate faulty results affecting the next time bucket.
In this repro case I create a CQ that runs sums on fields value50 and value60, then we insert a new point which introduces field value40. The results of the CQ generated from the newly inserted point are faulty.

**Repro steps**:

```
> create database t1

> use t1
Using database t1

> insert test value50=15,value60=20,proxy_count=10

> insert test value50=15,value60=20,proxy_count=10

> select * from test
name: test
----------
time				proxy_count	value50	value60
2016-10-07T13:46:25.434207216Z	10		15	20
2016-10-07T13:46:33.665756334Z	10		15	20


> create continuous query test1 on t1 resample for 1h begin select sum(value40)*sum(proxy_count) as mult40,sum(value50)*sum(proxy_count) as mult50,sum(value60)*sum(proxy_count) as mult60 into count from test group by time(1m) end

> select * from count
name: count
-----------
time			mult50	mult60
2016-10-07T13:46:00Z	600	800
```

The results are good so far.

The next time bucket however is generated with wrong results:

```
> insert test value40=10,value50=15,value60=20,proxy_count=10
> select * from test
name: test
----------
time				proxy_count	value40	value50	value60
2016-10-07T13:46:25.434207216Z	10			15	20
2016-10-07T13:46:33.665756334Z	10			15	20
2016-10-07T13:47:11.594310967Z	10		10	15	20


> select * from count
name: count
-----------
time			mult40	mult50	mult60
2016-10-07T13:46:00Z		600	800
2016-10-07T13:47:00Z	200	150	200
	
```

mult40 on 2016-10-07T13:47:00Z should be 10*10=100 instead of 200

Even though there is a resample clause that should force recalculations every minute over the last hour's data, the result is never corrected which probably means that the CQ still considers "200" as a correct result.

Manually executing the query generates correct results for the second time bucket:

```
> select sum(value40)*sum(proxy_count) as mult40,sum(value50)*sum(proxy_count) as mult50,sum(value60)*sum(proxy_count) as mult60 from test where time > now() - 1h group by time(1m)
time			mult40	mult50	mult60
...
2016-10-07T13:46:00Z		600	800
2016-10-07T13:47:00Z	100	150	200

```

Inserting the same values in a new time bucket (minute), the CQ generates correct results for it:

```
> insert test value40=10,value50=15,value60=20,proxy_count=10
> select * from test
name: test
----------
time				proxy_count	value40	value50	value60
2016-10-07T13:46:25.434207216Z	10			15	20
2016-10-07T13:46:33.665756334Z	10			15	20
2016-10-07T13:47:11.594310967Z	10		10	15	20
2016-10-07T13:56:18.114342318Z	10		10	15	20

> select * from count
name: count
-----------
time			mult40	mult50	mult60
2016-10-07T13:46:00Z		600	800
2016-10-07T13:47:00Z	200	150	200
2016-10-07T13:56:00Z	100	150	200

```

### PR DESCRIPTION
**Pull request by jsternberg**
Also fills in missing values using the fill expression for any binary
aggregation.

Fixes #7427 and #7469.

### CODE REVIEW

**Comment by jsternberg** [comment]
@mark-rushakoff this can act as an alternative to #7464 that also implements `fill()` for missing values. I'll also cherry pick the test case that you have in that PR so we can have it here too.

File: influxql/iterator.gen.go.tmpl [hunk 1]
```diff
@@ -1213,9 +1213,27 @@ func (itr *{{$k.name}}Stream{{$v.Name}}Iterator) reduce() ([]{{$v.Name}}Point, e
 // {{$k.name}}{{if ne $k.Name $v.Name}}{{$v.Name}}{{end}}ExprIterator executes a function to modify an existing point
 // for every output of the input iterator.
 type {{$k.name}}{{if ne $k.Name $v.Name}}{{$v.Name}}{{end}}ExprIterator struct {
-	left  *buf{{$k.Name}}Iterator
-	right *buf{{$k.Name}}Iterator
-	fn    {{$k.name}}{{if ne $k.Name $v.Name}}{{$v.Name}}{{end}}ExprFunc
+	left   *buf{{$k.Name}}Iterator
+	right  *buf{{$k.Name}}Iterator
+	fn     {{$k.name}}{{if ne $k.Name $v.Name}}{{$v.Name}}{{end}}ExprFunc
+	points []{{$k.Name}}Point // must be size 2
```
Comment (mark-rushakoff) [inline influxql/iterator.gen.go.tmpl]: Would it be better to have this be an array rather than a slice, so that the compiler can enforce length of 2?

Comment (jsternberg) [inline influxql/iterator.gen.go.tmpl]: I tried that at first, but then I can't set it to nil when `fill(none)` is specified. Oh, that reminds me. I forgot to implement `fill(previous)`....

**Comment by jsternberg** [comment]
Should this be the correct output of `fill(previous)`?

```
> select a, b, a + b from cpu fill(previous)
name: cpu
time                    a       b       a_b
----                    -       -       ---
1476729651246087386     1
1476729656794710992             2       3
1476729661978670263     3       4       7
```

**Comment by desa** [comment]
@jsternberg wouldn't we get something like
```
> select a, b, a + b from cpu fill(previous)
name: cpu
time                    a       b       a_b
----                    -       -       ---
1476729651246087386     1
1476729656794710992     1      2       3
1476729661978670263     3       4       7
```

or am I just miss understanding?

**Comment by jsternberg** [comment]
That's a good point, but I'm not sure I can do that as part of this PR since it's a separate part of the code. If we think that's a proper thing to do in the future, we might want to add the binary functionality now (like this PR) and then make a note in the docs that we're going to do it for the other types of queries too so it does the output that you said.

**Pull request merged**